File: src/main/java/gay/skitbet/jackiro/command/Command.java
package gay.skitbet.jackiro.command;

import gay.skitbet.jackiro.utils.JackiroModule;
import lombok.Data;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.interactions.commands.build.CommandData;
import net.dv8tion.jda.api.interactions.commands.build.OptionData;
import net.dv8tion.jda.internal.interactions.CommandDataImpl;

import java.util.ArrayList;
import java.util.List;

@Data
public abstract class Command {
    private final String name;
    private final String description;
    private final JackiroModule module;

    private Permission permission;

    public Command(String name, String description, JackiroModule module) {
        this.name = name.toLowerCase();
        this.description = description;
        this.module = module;
    }

    public Command(String name, String description, Permission permission, JackiroModule module) {
        this.name = name.toLowerCase();
        this.description = description;
        this.permission = permission;
        this.module = module;
    }

    public abstract void execute(CommandContext context);
    public abstract List<OptionData> addOptions(List<OptionData> options);

    public CommandData toData() {
        List<OptionData> options = addOptions(new ArrayList<>());

        return new CommandDataImpl(name, description)
                .addOptions(options);
    }
}

File: src/main/kotlin/gay/skitbet/jackiro/command/Command.kt
package gay.skitbet.jackiro.command

import gay.skitbet.jackiro.utils.JackiroModule
import lombok.Data
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.interactions.commands.build.CommandData
import net.dv8tion.jda.api.interactions.commands.build.OptionData
import net.dv8tion.jda.internal.interactions.CommandDataImpl
import java.util.*

@Data
abstract class Command {
    val name: String
    val description: String
    val module: JackiroModule?
    val ephemeral: Boolean

    var permission: Permission? = null

    constructor(name: String, description: String, module: JackiroModule?, ephemeral: Boolean) {
        this.name = name.lowercase(Locale.getDefault())
        this.description = description
        this.module = module
        this.ephemeral = ephemeral
    }

    constructor(
        name: String,
        description: String,
        permission: Permission?,
        module: JackiroModule?,
        ephemeral: Boolean
    ) {
        this.name = name.lowercase(Locale.getDefault())
        this.description = description
        this.permission = permission
        this.module = module
        this.ephemeral = ephemeral
    }

    abstract fun execute(context: CommandContext)
    abstract fun addOptions(options: MutableList<OptionData?>): MutableList<OptionData?>

    fun toData(): CommandData {
        val options = addOptions(ArrayList<OptionData?>())

        return CommandDataImpl(name, description)
            .addOptions(options)
    }
}
